mllib/src/main/scala/org/apache/spark/mllib/regression/GeneralizedLinearAlgorithm.scala

```diff
@@ -179,7 +179,7 @@ abstract class GeneralizedLinearAlgorithm[M <: GeneralizedLinearModel]
    */
   def run(input: RDD[LabeledPoint]): M = {
     if (numFeatures < 0) {
-      numFeatures = input.map(_.features.size).first()
+      numFeatures = input.first().features.size
```

Comment:
It would only evaluate one element in this case. first() accesses the first partition and first element of its iterator, so map should only be applied once.